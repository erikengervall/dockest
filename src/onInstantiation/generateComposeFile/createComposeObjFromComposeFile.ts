import { ComposeFile } from '../../runners/@types'
import mergeDockerComposeConfigFiles from './mergeDockerComposeConfigFiles'

export default (composeFiles: string[], nodeProcess = process): ComposeFile => {
  const cwd = nodeProcess.cwd()

  let composeObj = {
    version: '3.7',
    services: {},
  }

  if (composeFiles.length > 0) {
    composeObj = mergeDockerComposeConfigFiles(cwd, composeFiles)
  }

  return composeObj
}
