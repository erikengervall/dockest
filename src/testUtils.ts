import { createMockProxy } from 'jest-mock-proxy'
import execa from 'execa' // eslint-disable-line import/default
import { ZooKeeperRunner, KafkaRunner, PostgresRunner, RedisRunner, GeneralPurposeRunner } from './runners'
import { Runner } from './runners/@types'
import Logger from './Logger'
import { DEFAULT_USER_CONFIG, INTERNAL_CONFIG } from './constants'

export const mockedExecaStdout = 'getContainerId ðŸŒ®'
export const runnerCommand = 'runRunnerCommands ðŸŒ®'

/**
 * FIXME: Would love for this to work ðŸ¦Š
 */
// jest.mock('execa', () => jest.fn(() => ({ stdout: mockedExecaStdout })))
// jest.mock('./Logger')

const { values } = Object

const createDockestConfig = (opts: { runners?: Runner[] }) => {
  const { runners = [] } = opts

  for (const runner of runners) {
    runner.logger = createMockProxy()
  }

  return {
    runners,
    opts: DEFAULT_USER_CONFIG,
    jest: {},
    $: INTERNAL_CONFIG,
  }
}

export default ({ withRunnerCommands = false }) => {
  const withCmds = withRunnerCommands ? { commands: [runnerCommand] } : {}

  const zooKeeperRunner = new ZooKeeperRunner({ service: 'zookeepeer', image: 'some/image:123', ...withCmds })
  const kafkaRunner = new KafkaRunner({
    service: 'kafka',
    image: 'some/image:123',
    dependsOn: [zooKeeperRunner],
    ...withCmds,
  })
  const postgresRunner = new PostgresRunner({
    service: 'postgres',
    image: 'some/image:123',
    database: '_',
    username: '_',
    password: '_',
    ...withCmds,
  })
  const redisRunner = new RedisRunner({ service: 'redis', image: 'some/image:123', ...withCmds })
  const generalPurposeRunner = new GeneralPurposeRunner({ service: 'general', image: 'some/image:123', ...withCmds })
  const initializedRunners = {
    kafkaRunner,
    postgresRunner,
    redisRunner,
    zooKeeperRunner,
    generalPurposeRunner,
  }

  beforeEach(() => values(initializedRunners).forEach(runner => (runner.logger = createMockProxy())))

  return {
    initializedRunners,
    runners: {
      ZooKeeperRunner,
      KafkaRunner,
      PostgresRunner,
      RedisRunner,
      GeneralPurposeRunner,
    },
    execa,
    Logger,
    createDockestConfig,
  }
}
