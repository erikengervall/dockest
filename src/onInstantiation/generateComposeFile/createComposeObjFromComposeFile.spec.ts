import fs from 'fs'
import yaml from 'js-yaml'
import createComposeObjFromComposeFile from './createComposeObjFromComposeFile'
import testUtils from '../../testUtils'

const { createDockestConfig } = testUtils({})

describe('createComposeObjFromComposeFile', () => {
  const yamlMock: any = {
    safeLoad: jest.fn(composeYml => yaml.safeLoad(composeYml)),
  }
  const fsMock: any = {
    readFileSync: jest.fn((filePath, encoding) => fs.readFileSync(`${process.cwd()}${filePath}`, encoding)),
  }
  const nodeProcessMock: any = {
    cwd: jest.fn(() => '/fixtures/'),
  }

  beforeEach(() => {
    yamlMock.safeLoad.mockClear()
    fsMock.readFileSync.mockClear()
    nodeProcessMock.cwd.mockClear()
  })

  it('should create composeObj from single service Compose File', () => {
    const dockestConfig = createDockestConfig({
      opts: { composeFile: 'docker-compose-single-redis.yml' },
    })

    const composeObj = createComposeObjFromComposeFile(dockestConfig, fsMock, yamlMock, nodeProcessMock)

    expect(composeObj).toMatchSnapshot()
  })

  it('should create composeObj from complicated Compose File', () => {
    const dockestConfig = createDockestConfig({
      opts: { composeFile: 'docker-compose-complicated.yml' },
    })

    const composeObj = createComposeObjFromComposeFile(dockestConfig, fsMock, yamlMock, nodeProcessMock)

    expect(composeObj).toMatchSnapshot()
  })

  describe('multiple Compose Files', () => {
    it('should create composeObj from multiple Compose Files', () => {
      const dockestConfig = createDockestConfig({
        opts: { composeFile: ['docker-compose-single-redis.yml', 'docker-compose-single-postgres.yml'] },
      })

      const composeObj = createComposeObjFromComposeFile(dockestConfig, fsMock, yamlMock, nodeProcessMock)

      expect(composeObj).toMatchSnapshot()
    })

    it('should merge conflicting names', () => {
      const dockestConfig = createDockestConfig({
        opts: { composeFile: ['docker-compose-single-redis.yml', 'docker-compose-single-redis-duplicate.yml'] },
      })

      const composeObj = createComposeObjFromComposeFile(dockestConfig, fsMock, yamlMock, nodeProcessMock)

      expect(composeObj).toMatchSnapshot()
    })
  })
})
