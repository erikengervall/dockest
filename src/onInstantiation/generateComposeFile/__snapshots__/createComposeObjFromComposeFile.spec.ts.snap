// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createComposeObjFromComposeFile multiple Compose Files should create composeObj from multiple Compose Files 1`] = `
Object {
  "services": Object {
    "composePostgres": Object {
      "environment": Object {
        "POSTGRES_DB": "nobueno",
        "POSTGRES_PASSWORD": "is",
        "POSTGRES_USER": "ramda",
      },
      "image": "postgres:9.6",
      "ports": Array [
        Object {
          "published": 5433,
          "target": 5432,
        },
      ],
    },
    "composeRedis": Object {
      "image": "redis:5.0.3",
      "ports": Array [
        Object {
          "published": 6378,
          "target": 6379,
        },
      ],
    },
  },
  "version": "3.7",
}
`;

exports[`createComposeObjFromComposeFile multiple Compose Files should merge conflicting names 1`] = `
Object {
  "services": Object {
    "composeRedis": Object {
      "image": "im:a:duplicate",
      "ports": Array [
        Object {
          "published": 6378,
          "target": 6379,
        },
      ],
    },
  },
  "version": "3.7",
}
`;

exports[`createComposeObjFromComposeFile should create composeObj from complicated Compose File 1`] = `
Object {
  "services": Object {
    "dockest_inline_service_name_postgres1sequelizeRunner": Object {
      "environment": Object {
        "POSTGRES_DB": "nobueno",
        "POSTGRES_PASSWORD": "is",
        "POSTGRES_USER": "ramda",
      },
      "image": "postgres:9.6",
      "ports": Array [
        Object {
          "published": 5433,
          "target": 5432,
        },
      ],
    },
    "dockest_kafka1confluentinc": Object {
      "depends_on": Array [
        "dockest_zookeeper1confluentinc",
      ],
      "environment": Object {
        "KAFKA_ADVERTISED_LISTENERS": "PLAINTEXT://dockest_kafka1confluentinc:29092,PLAINTEXT_HOST://localhost:9092",
        "KAFKA_AUTO_CREATE_TOPICS_ENABLE": "true",
        "KAFKA_BROKER_ID": 1,
        "KAFKA_LISTENER_SECURITY_PROTOCOL_MAP": "PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT",
        "KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR": 1,
        "KAFKA_ZOOKEEPER_CONNECT": "dockest_zookeeper1confluentinc:2181",
      },
      "image": "confluentinc/cp-kafka:5.2.2",
      "ports": Array [
        Object {
          "published": 9092,
          "target": 9092,
        },
      ],
    },
    "dockest_postgres2knex": Object {
      "environment": Object {
        "POSTGRES_DB": "nobueno",
        "POSTGRES_PASSWORD": "is",
        "POSTGRES_USER": "ramda",
      },
      "image": "postgres:9.6",
      "ports": Array [
        Object {
          "published": 5434,
          "target": 5432,
        },
      ],
    },
    "dockest_redis1ioredis": Object {
      "image": "redis:5.0.3",
      "ports": Array [
        Object {
          "published": 6378,
          "target": 6379,
        },
      ],
    },
    "dockest_zookeeper1confluentinc": Object {
      "environment": Object {
        "ZOOKEEPER_CLIENT_PORT": "2181",
      },
      "image": "confluentinc/cp-zookeeper:5.2.2",
      "ports": Array [
        Object {
          "published": 2181,
          "target": 2181,
        },
      ],
    },
  },
  "version": "3.7",
}
`;

exports[`createComposeObjFromComposeFile should create composeObj from single service Compose File 1`] = `
Object {
  "services": Object {
    "composeRedis": Object {
      "image": "redis:5.0.3",
      "ports": Array [
        Object {
          "published": 6378,
          "target": 6379,
        },
      ],
    },
  },
  "version": "3.7",
}
`;
